File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque
N, line, s = list(map(int, input().split()))

arr = [[0 for _ in range(N+1)] for _ in range(N+1)]

for i in range(line):
    x, y = map(int, input().split())
    arr[x][y] = 1
    arr[y][x] = 1

def bfs(start):
    deq = deque()
    visited = [0 for i in range(N+1)]
    visited[start] = 1
    deq.append(start)
    while len(deq)!=0:
        current = deq.popleft()
        print(current, end=' ')

        for i in range(1, N+1):
            if(arr[current][i]==1 and visited[i]==0):
                visited[i] = 1
                deq.append(i)

visited1 = [False] * (N+1)  # dfs의 방문기록

def dfs(start):
    visited1[start] = 1 
    print(start, end=" ")
    for i in range(1, N+1):
        if visited1[i]==0 and arr[start][i]==1:  
            dfs(i)  


dfs(s)
print()
bfs(s)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : 'input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const startNode = Number(input[0].split(' ')[2]);
const graph = input.slice(1).map((value) => value.split(' ').map(Number));

function bfs(graph, startNode) {
  const visited = new Set();
  const queue = [];
  const result = [];
  let front = 0;

  queue.push(startNode);
  visited.add(startNode);

  while (front < queue.length) {
    const currentNode = queue[front++];
    result.push(currentNode);

    const neighbors = graph[currentNode] || [];
    for (const neighbor of neighbors) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        queue.push(neighbor);
      }
    }
  }

  console.log(result.join(' '));
}

function dfs(graph, startNode) {
  const stack = [startNode];
  const visited = new Set();
  const result = [];

  while (stack.length > 0) {
    const currentNode = stack.pop();

    if (!visited.has(currentNode)) {
      result.push(currentNode);
      visited.add(currentNode);

      const neighbors = graph[currentNode] || [];
      for (const neighbor of neighbors.reverse()) {
        stack.push(neighbor);
      }
    }
  }

  console.log(result.join(' '));
}

function createGraph(input) {
  const graph = {};

  for (const [node1, node2] of input) {
    if (!graph[node1]) {
      graph[node1] = [];
    }
    if (!graph[node2]) {
      graph[node2] = [];
    }

    graph[node1].push(node2);
    graph[node2].push(node1);

    graph[node1].sort((a, b) => a - b);
    graph[node2].sort((a, b) => a - b);
  }

  return graph;
}

dfs(createGraph(graph), startNode);
bfs(createGraph(graph), startNode);
